src/rgen/Microsoft.Macios.Generator/TabbedStringBuilder.cs

```diff
@@ -0,0 +1,188 @@
+using System;
+using System.Text;
+
+namespace Microsoft.Macios.Generator;
+
+
+/// <summary>
+/// String builder wrapper that keeps track of the current indentation level by abusing the IDisposable pattern. Rather
+/// than dispose data, what the IDisposable pattern allows is to create a new block with an increased indentation, that
+/// way we do not need to keep track of the current indentation level.
+/// <example>
+/// var classBlock = new TabbedStringBuilder (sb);
+/// classBlock.AppendFormatLine ("public static NSString? GetConstant (this {0} self)", enumSymbol.Name);
+/// // open a new {} block, no need to keep track of the indentation, the new block has it
+/// using (var getConstantBlock = classBlock.CreateBlock (isBlock: true)) {
+///    // write the contents of the method here.
+/// }
+/// </example>
+/// </summary>
+class TabbedStringBuilder : IDisposable {
+	readonly StringBuilder sb;
+	readonly uint tabCount;
+	readonly string tabs;
+	readonly bool isBlock;
+	bool disposed;
+
+	/// <summary>
+	/// Created a new tabbed string builder that will use the given sb to write code.
+	/// </summary>
+	/// <param name="builder">The string builder to be used to write code.</param>
+	/// <param name="currentCount">The original tab size.</param>
+	/// <param name="block">States if we are creating a {} block.</param>
+	public TabbedStringBuilder (StringBuilder builder, uint currentCount = 0, bool block = false)
+	{
+		sb = builder;
+		isBlock = block;
+		if (isBlock) {
+			// increase by 1 because we are in a block
+			tabCount = currentCount + 1;
+			var braceTab = new string ('\t', (int) (tabCount - 1));
+			this.sb.AppendLine ($"{braceTab}{{");
+		} else {
+			tabCount = currentCount;
+		}
+		tabs = new string ('\t', (int) tabCount);
```

Comment:
Instead of creating and writing a string, I think it would be better to have a WriteTabs method:

```cs
void WriteTabs ()
{
    sb.Append ('\t', tabCount);
}
```

and then call that instead of writing the string.

At the very least we'll have fewer string allocations (which is always good).